cpupower_monitor.py

```diff
@@ -0,0 +1,115 @@
+            if 'Available idle states: ' in line:
+                self.states_list = line.split()
+        del self.states_list[0:3]
+        self.log.info("Idle states on the system are: ", self.states_list)
+        self.run_cmd_out("cpupower monitor")
+
+    def run_cmd_out(self, cmd):
+        return process.system_output(cmd, shell=True, ignore_status=True,
+                                     sudo=True).decode("utf-8")
+
+    def check_zero_nonzero(self, cmd, stop_state_index):
+        self.output = self.run_cmd_out(cmd)
+        values_list = []
+        for line in self.output.splitlines():
+            stop_state_values = line.split('|')
+            values_list.append(stop_state_values[stop_state_index].strip())
+        for value in values_list:
+            if float(value) != 0.00:
+                return 1
+        return 0
+
+    def test_workload(self):
+        tarball = self.fetch_asset('http://sourceforge.net/projects/ebizzy/files/ebizzy/' \
+                                   '0.3/ebizzy-0.3.tar.gz')
+        archive.extract(tarball, self.workdir)
+        self.sourcedir = os.path.join(self.workdir, 'ebizzy-0.3')
+        os.chdir(self.sourcedir)
+        process.run('./configure', shell=True)
+        build.make(self.sourcedir)
+        self.run_cmd_out("cpupower monitor")
+        self.log.info("============ Starting ebizzy tests ==============")
+        obj = process.SubProcess('./ebizzy &', verbose=False,
+                                         shell=True)
+        obj.start()
+        time.sleep(2)
+        for i in range(len(self.states_list)):
+            zero_nonzero = self.check_zero_nonzero("cpupower monitor | grep -v '[A-Z]'", i + 1)
+            if zero_nonzero:
+                self.fail("cpus entered idle states during ebizzy workload")
+            else:
+                self.log.info("no cpus entered idle states while running ebizzy")
+        time.sleep(10)
+        zero_nonzero = 0
+        for i in range(len(self.states_list)):
+            zero_nonzero = zero_nonzero + self.check_zero_nonzero("cpupower monitor | grep -v '[A-Z]'", i + 1)
+        if zero_nonzero:
+            self.log.info("cpus have entered idle states after killing work load")
+        else:
+            self.fail("cpus have not entered idle states after killing ebizzy workload")
+
+    def test_disable_idlestate(self):
+        for i in range(len(self.states_list)):
+            process.run('cpupower -c all idle-set -d %s' % i, shell=True)
+            time.sleep(5)
+            zero_nonzero = self.check_zero_nonzero("cpupower monitor | grep -v '[A-Z]'", i + 1)
+            if zero_nonzero:
+                self.fail("cpus have entered the disabled idle states.")
+            else:
+                self.log.info("cpus have not entered disabled idle states")
+            process.run('cpupower -c all idle-set -E', shell=True)
+
+    def test_idlestate_smt(self):
+        process.run('ppc64_cpu --smt=off', shell=True)
```

Comment:
Testcases like this specific to Power arch can be skipped on other architectures.